src/osd/PrimaryLogPG.cc

```diff
@@ -1529,6 +1529,47 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
   return r;
 }
 
+// it is pretty much the same as CEPH_OSD_OP_COPY_FROM, but it requires user to
+// specify the src osd, and uses repair_read() to read the object payload, also
+// the src oid is always identical to the dest oid (the one in OpContext)
+int PrimaryLogPG::do_repair_copy(OpContext *ctx, const OSDOp& osd_op, bufferlist::iterator& bp)
+{
+  repair_copy_arg_t arg;
+
+  vector<pg_shard_t> bad_shards;
+  try {
+    arg.decode(bp);
+  } catch (buffer::error&) {
+    dout(10) << "corrupted " << __func__ << " arg" << dendl;
+    return -EINVAL;
+  }
+  const auto& oi = ctx->obs->oi;
+  tracepoint(osd, do_osd_op_pre_repair_copy,
+	     oi.soid.oid.name.c_str(),
+	     oi.soid.snap.val,
+	     arg.what);
+
+  if (arg.bad_shards.empty() || !arg.what) {
+    return 0;
+  }
+  if (!ctx->copy_cb) {
+    // mark the bad shard missing so they are not chosen when repairing
+    for (const auto& bad_shard : arg.bad_shards) {
+      peer_missing[bad_shard].add(oi.soid, oi.version, eversion_t());
```

Comment:
i did investigate the possible ways to pass this down. it's possible, but complicates the code path a lot. marking the corrupted shard missing fits our needs pretty well.

> I'm worried one of those would be affected by adding temporary entries to peer_missing.

i will investigate this. and come back to you.